### PR TITLE
http: deprecate more of coding infrastructure

### PR DESCRIPTION
**Pull request by jrudolph**
Refs #2916

Previously, we exposed Scala objects as API directly which exposes implementation
details and makes it hard to evolve the implementation. Instances in coders
are upcasted to `Coder` so that implementation details are not leaked any more.

### CODE REVIEW

**Comment by ennru** [comment]
The doc examples show `Gzip.withLevel(9)` which I silenced for now.

**Comment by jrudolph** [comment]
> The doc examples show `Gzip.withLevel(9)` which I silenced for now.

Thanks, I tried to go through the warnings, but it seems adding a deprecation didn't recompile dependencies so I missed all of those.

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Coders.scala [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import com.github.ghik.silencer.silent
+
+import scala.collection.immutable
+
+@silent("in package coding is deprecated")
+object Coders {
+  def Gzip: Coder = akka.http.scaladsl.coding.Gzip
+  def Gzip(messageFilter: HttpMessage => Boolean): Coder = new Gzip(messageFilter)
+
+  def Deflate: Coder = akka.http.scaladsl.coding.Gzip
```
Comment (jrudolph) [inline akka-http/src/main/scala/akka/http/scaladsl/coding/Coders.scala]: oops

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Coders.scala [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package akka.http.scaladsl.coding
+
+import akka.http.scaladsl.model.HttpMessage
+import com.github.ghik.silencer.silent
+
+import scala.collection.immutable
+
+@silent("in package coding is deprecated")
+object Coders {
```
[the remark this answers is not in the document]
Comment (raboof) [reply akka-http/src/main/scala/akka/http/scaladsl/coding/Coders.scala:13]: Good to be consistent with e.g. `StatusCodes` I think, indeed

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Gzip.scala [hunk 3]
```diff
@@ -4,10 +4,13 @@
 
 package akka.http.scaladsl.coding
 
+import akka.annotation.InternalApi
 import akka.http.scaladsl.model._
 import akka.http.scaladsl.model.headers.HttpEncodings
 
-class Gzip private (compressionLevel: Int, val messageFilter: HttpMessage => Boolean) extends Coder with StreamDecoder {
+@InternalApi
+@deprecated("Actual implementation of Gzip is internal, use Coders.Gzip instead", since = "10.2.0")
```
Comment (raboof) [inline akka-http/src/main/scala/akka/http/scaladsl/coding/Gzip.scala:12]: It's a bit strange to mark internal API as deprecated, but indeed as it used to be public API I guess it makes sense :+1:

**Comment by ennru** [comment]
I did another round on this to use the non-deprecated variants in the tests.

File: akka-http/src/main/scala/akka/http/scaladsl/coding/Gzip.scala (hunk 3 above)
Comment (jrudolph) [reply akka-http/src/main/scala/akka/http/scaladsl/coding/Gzip.scala:12]: Yeah, guess I went over the top with deprecating stuff, just to be sure. Maybe we can create an internal copy later as the future version while keeping that one around only for compatibility.

**Comment by jrudolph** [comment]
> I did another round on this to use the non-deprecated variants in the tests.

Thanks, much appreciated :)

I'm planning to do a more thorough review of warnings and documentation for 10.2.0 next week.

**Pull request merged**